elastic/kibana
# [Monitoring] Fix inaccuracies in logstash pipeline listing metrics

Resolves #55755

@inqueue noticed an issue with what shows up in the Logstash Overview Events Emitted chart versus what shows in the Logstash Pipelines Listing Events Emitted listing.

In the overview chart, you can see the total events emitted are ~77k.
<img width="1146" alt="Screen Shot 2020-01-24 at 11 29 44 AM" src="https://user-images.githubusercontent.com/56682/73085668-0f8e6800-3e9d-11ea-979b-1059c0bde051.png">

In the listing page, if you sum the events emitted, it's only ~6k
<img width="539" alt="Screen Shot 2020-01-24 at 11 30 49 AM" src="https://user-images.githubusercontent.com/56682/73085669-0f8e6800-3e9d-11ea-9973-69af27c63199.png">

After some digging (and help from @ycombinator), we discovered a couple of things:

- https://github.com/elastic/kibana/pull/49978 introduced a regression where we'd not capture the entire list of `ephemeral_id` and `hash` values we needed to calculate the right metrics
- The way we calculated the data differed enough to still cause inconsistencies after we fixed the above.

This PR addresses that by ensuring we execute the same kind of query so we get the same results. We are able to do that by, instead of sending a single query to fetch all pipeline metrics, we send individual ones (in parallel) for each known pipeline. 

Now, we see:
<img width="511" alt="Screen Shot 2020-01-24 at 12 45 44 PM" src="https://user-images.githubusercontent.com/56682/73091016-86306300-3ea7-11ea-86ae-c5673135309e.png">


It's a little complicated with logstash monitoring documents, due to the nested fields, so you'll notice some of that in the code.

## Testing

- Setup logstash monitoring and ensure logstash is running with multiple pipelines
- Verify the events emitted chart on the overview page is indeed the sum of the events emitted metrics on the listing page
